benchmarks/Lua/som.lua

```diff
@@ -0,0 +1,631 @@
+
+        if n > 3 then           -- More than three elements.
+            -- Find k>i and l<j such that dk,dij,dl are in reverse order.
+            -- Swap k and l.  Repeat this procedure until k and l pass each other.
+            local k = i
+            local l = j - 1
+
+            while true do
+                -- i.e. while dl succeeds dij
+                while k <= l and fn(dij, storage[l]) do
+                    l = l - 1
+                end
+
+                k = k + 1
+                -- i.e. while dij succeeds dk
+                while k <= l and fn(storage[k], dij) do
+                    k = k + 1
+                end
+
+                if k > l then
+                    break
+                end
+
+                local tmp = storage[k]
+                storage[k] = storage[l]
+                storage[l] = tmp
+            end
+
+            -- Now l < k (either 1 or 2 less), and di through dl are all
+            -- less than or equal to dk through dj.  Sort those two segments.
+            self:sort_range(i, l, fn)
+            self:sort_range(k, j, fn)
+        end
+    end
+end
+
+end -- class Vector
+
+local Set = {_CLASS = 'Set'} do
+
+local INITIAL_SIZE = 10
+
+function Set.new (size)
+    local obj = {
+        items = Vector.new(size or INITIAL_SIZE)
+    }
+    return setmetatable(obj, {__index = Set})
+end
+
+function Set:size ()
+    return self.items:size()
+end
+
+function Set:each ()
+    return self.items:each()
+end
+
+function Set:has_some (fn)
+    return self.items:has_some(fn)
+end
+
+function Set:get_one (fn)
+    return self.items:get_one(fn)
+end
+
+function Set:add (obj)
+    if not self:contains(obj) then
+        self.items:append(obj)
+    end
+end
+
+function Set:remove_all ()
+    self.items:remove_all()
+end
+
+function Set:collect (fn)
+    local coll = Vector.new()
+    for e in self:each() do
+        coll:append(fn(e))
+    end
+    return coll
+end
+
+function Set:contains (obj)
+    return self:has_some(function (it) return it == obj end)
+end
+
+end -- class Set
+
+local IdentitySet = {_CLASS = 'IdentitySet'} do
+setmetatable(IdentitySet, {__index = Set})
+
+function IdentitySet.new (size)
+    local obj = Set.new(size)
+    return setmetatable(obj, {__index = IdentitySet})
+end
+
+function IdentitySet:contains (obj)
+    return self:has_some(function (it) return it == obj end)
+end
+
+end -- class IdentitySet
+
+local Entry = {_CLASS = 'Entry'} do
+
+function Entry.new (hash, key, value, next)
+    local obj = {
+        hash  = hash,
+        key   = key,
+        value = value,
+        next  = next,
+    }
+    return setmetatable(obj, {__index = Entry})
+end
+
+function Entry:match (hash, key)
+    return self.hash == hash and self.key == key
+end
+
+end -- class Entry
+
+local Dictionary = {_CLASS = 'Dictionary'} do
+
+local INITIAL_CAPACITY = 16
+local floor = math.floor
+
+function Dictionary.new (size)
+    local obj = {
+        buckets = {n = size or INITIAL_CAPACITY},
```

Comment:
for the dictionary storage/buckets array, we also should try to make sure that it is a continuous array. Otherwise, performance for DeltaBlue and Havlak, I think, is likely very bad.